Fix: construction of query string for EDS Autocomplete requests

cwolfebsco opened the pull request:
Current implementation of EDS Autocomplete URL fails to resolve correctly and Autocomplete did not work. This minor fix introduces an array of required parameters and use's PHP's built in `http_build_query` to correctly built the URL.

Example of Autocomplete URL that was built by original function:

`https://global.ac.ebsco-content.com/autocomplete/rest/autoComplete?idx=rawqueries&token=AhC%2Bjs6yALBfPr7DbjKT5r0ASLTpEESVpsbUcdq%2BdsZE4xMDZEHXZQylrftQvsOAr0nrkUZ0AA%3D%3D&filters=[{"name"%3A"custid"%2C"values"%3A["ns123456"]}]&term=bacter`

Example of Autocomplete URL that was built by amended function:

`https://global.ac.ebsco-content.com/autocomplete/rest/autoComplete?idx=rawqueries&token=AhDQQ4bnYjJIjTTXxaBdpyIMwNifUA%2FotBOvrMOmJD0ckVos4IolahosrYstNMYdYBIpY8SuxA%3D%3D&filters=%5B%7B%22name%22%3A%22custid%22%2C%22values%22%3A%5B%22ns123456%22%5D%7D%5D&term=bacter`

Note: customer id was obfuscated

cwolfebsco committed:
Fix: Construction of Query String for EDS Autocomplete Requests

EDS autocomplete fails within the old code as the json that was enclosed in the filters parameters wasn't fully URLEncoded.

This changes create an array of parameters and then uses a built in PHP function to correctly build the query string. Testing with this amended function worked correctly locally and for a German VuFind User

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -269,11 +269,14 @@ protected function parseAutocomplete($msg)
+        $params = [];
+        $params['idx'] = $type;
+        $params['token'] = $data['token'];
+        $params['filters'] = '[{"name" : "custid", "values": ["'.$data['custid'].'"]}]';
```
Might it be better to change:

`$params['filters'] = '[{"name" : "custid", "values": ["'.$data['custid'].'"]}]';`

to:

`$params['filters'] = '[{"name" : "custid", "values": ' . json_encode([$data['custid']]) . '}]';`

...to ensure that the custid value is always appropriately encoded?

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Backend/EDS/Base.php:
```diff
@@ -269,11 +269,14 @@ protected function parseAutocomplete($msg)
+        $params['filters'] = '[{"name" : "custid", "values": ["'.$data['custid'].'"]}]';
+        $params['term'] = $query;
+        
+        $url = $data['url'].'?'.http_build_query($params);
```
I suspect that this is going to cause the build to fail on style grounds. I'd recommend running `vendor/bin/phing php-cs-fixer` to correct style issues. You can also `composer run qa` to make sure all of the checks will pass before pushing things up. I'm happy to help resolve any issues if something fails in an unclear way.

cwolfebsco replied:
Thanks for the generous feedback - I might need to take you up on some more guidance.

For the filter's object I decided to go ahead and build the array of objects more explicitly and then json_encode the whole thing: https://github.com/vufind-org/vufind/pull/2319/commits/d0272124c3954fd5cae0c56b42802acd2074773d

I tried follow instructions on https://vufind.org/wiki/development:recommended_tools for vendor/bin/phpcs --standard=PEAR [file] and hope that helps the checks

I must look into composer run qa - it throws an error at my end :( => Script "qa" is not defined in this package  ... I am sure that is just me not following the steps quite right

demiankatz replied:
Regarding the filter construction, this might be a slightly more concise approach to building the data structure:

<pre>
    $filters = [['name' => 'custid', 'values' => [$data['custid']]];
</pre>

Regarding `composer run qa`, is it possible you're working with an older version of VuFind? It was added relatively recently, so perhaps you just haven't pulled the latest dev code yet. It should be defined here:

https://github.com/vufind-org/vufind/blob/dev/composer.json#L129

cwolfebsco replied:
I am so sorry for making your life this hard for you  and thank you for humouring my beginner's mistakes.

I have adapted your suggested filter construction (I always get lost with all those square brackets, this one needed four closing ones 🤯, but it is definitely a lot nicer than my quick one.

I also added a unit test that will correctly assert the changed structure.

I ended up checking out the latest dev branch and got through most of running the composer run qa - it helped me find some more stuff, though it did fail... I might spend some more time on it on Monday

demiankatz replied:
@cwolfebsco, no need to apologize -- I'm really more than happy to help, and appreciate your willingness to respond to my feedback! Always happy to invest a little time in on-boarding, because it usually pays off in the long term.

I should note that I'll be on vacation for most of next week (out of office from 2/15-2/21), so I may be less responsive than usual during that period... but I'll be happy to continue to answer questions as they arise, though I may not be in a position to finalize this before I get back.

demiankatz replied:
...and look at that: the build just passed! I think this is quite possibly in a position to merge, but please let me know if you have more work you want to do before that happens.

cwolfebsco replied:
I hope you enjoy your vacation! I did run `composer run qa` again this morning and it exists after a timeout, but this seems to be more an issue of the resource constraint VM I am using. If you are happy with this one, I'd be happy to see it merged :)

demiankatz replied:
Thanks! I made a few last-minute tweaks (it looks like some of the style changes applied to the comments weren't necessary based on VuFind project standards, so I reverted them to keep the commit simple, and I reformatted the array construction a little more based on personal preferences, though that's more a matter of taste than anything). Merging now!

demiankatz reviewed (changes requested):
Thanks for this fix! Just a couple of questions/comments. Happy to assist further if I can!

cwolfebsco committed:
Fix: Enhance structure of function

Create filters in a more structured way and json_encode the entire array of objects

cwolfebsco committed:
Update: whitespace updates for phpcs checks

cwolfebsco committed:
Fix: More concise filter creation statement & whitespace cleanup

Implemented a more precise filters array of objects variable, as suggested by Damien / cleaned up Whitespace

cwolfebsco committed:
Fix: Updated Filter to correctly reflect changes to function

Function now properly URLencodes the entire JSON object, this is represented in the changed test

demiankatz committed:
Fix style of comments.

demiankatz committed:
Build array in one step.

demiankatz committed:
Remove blank line.
